x/migration/keeper/msg_server_claim_morse_supplier.go

```diff
@@ -29,6 +29,28 @@ func (k msgServer) ClaimMorseSupplier(
 	sdkCtx := cosmostypes.UnwrapSDKContext(ctx)
 	logger := k.Logger().With("method", "ClaimMorseSupplier")
 
+	// Ensure that gas fees are NOT waived if the claim is invalid.
+	// This restores the disincentive for Shannon account holders to spam invalid
+	// Morse claim txs BUT ONLY if the spanner is using a funded account.
+	// In cases where the tx contains ONLY one or more Morse claim messages, the
+	// validators which consider the tx WILL do slightly more work than the typical
+	// message, which would've been rejected prior to the message handler during CheckTx.
+	var (
+		morseClaimableAccount              migrationtypes.MorseClaimableAccount
+		isFound, isValid, isAlreadyClaimed bool
+	)
+	defer func() {
+		if !isFound || !isValid || isAlreadyClaimed {
+			// Attempt to charge the waived gas fee for invalid claims.
+			sdkCtx.GasMeter()
+			// DEV_NOTE: Assuming that the tx containing this message was signed
```

Comment:
The assumption is fine. 

Any way we can check it just to make sure `x/` code errors rather than panic in the offchance.

Reply:
1. There is no way to check AFAIK because we cannot get a reference to the tx in this scope (message handler). Typically the ante handler deals with this long before any message handlers are called.
2. What makes you say that there would be a panic? The consequence of this assumption being wrong is NOT a panic, rather it's charging the incorrect gas amount for that key type. For additional context, the gas amount for each key type is its own `auth` module param.

Reply:
I've also iterated a bit on the ante handler logic to restrict the gas waiver to txs with exactly one secp256k1 signer. This hardens things a bit further.